enhancements/cluster-api/crd-compatibility-checker.md

```diff
@@ -0,0 +1,469 @@
+CCAPIO scans these, but does not load them.
+For each CRD it creates a new CRDCompatibilityRequirement.
+e.g. for the `Cluster` CRD it would create:
+
+```yaml
+apiVersion: operator.openshift.io/v1alpha1
+kind: CRDCompatibilityRequirement
+metadata:
+  name: cluster-api-cluster-ccapio-4.Y+1
+spec:
+  crdRef: clusters.cluster.x-k8s.io
+  creatorDescription: "OpenShift Cluster CAPI Operator for OCP 4.Y+1"
+  compatibilityCRD: |
+    ...
+    <complete YAML document of future Cluster CRD from transport config map>
+    ...
+  crdAdmitAction: Warn
+```
+
+The differences from the previous example are:
+* The requirement identifies itself as being related to a different version.
+* `crdAdmitAction` is `Warn` instead of `Enforce`, so this requirement will not prevent updates to the current CRD.
+* There is no `objectSchemaValidation`.
+
+CCAPIO will mark itself as not upgradeable if any CRDCompatibilityRequirement created for an upgrade transport config map reports its `Compatible` condition as `False`.
+
+### API Extensions
+
+The following is an outline of the CRDCompatibilityRequirement CRD:
+
+```yaml
+apiVersion: operator.openshift.io/v1alpha1
+kind: CRDCompatibilityRequirement
+metadata:
+  ...
+spec:
+  # Fields to be implemented initially
+  crdRef: <the name of the target CRD>
+  creatorDescription: |
+    <human readable description of who created this requirement. displayed in errors and warnings the validation webhooks>
+  compatibilityCRD: |
+    <the complete CRD required by the creator of this requirement, as yaml>
+  crdAdmitAction: <Enforce or Warn>
+
+  # If defined, also validate that matching objects conform to the schema given in compatibilityCRD.
+  # This will be implemented in a second phase. The initial API will not contain this field.
+  objectSchemaValidation:
+    # namespaceSelector, objectSelector, and matchConditions will be copied to
+    # the corresponding ValidatingWebhookConfiguration. Their definitions and
+    # semantics are therefore identical to those in
+    # ValidatingWebhookConfiguration.
+    namespaceSelector:
+      matchLabels:
+        "kubernetes.io/metadata.name": "openshift-cluster-api"
+    objectSelector:
+      ...
+    matchConditions:
+      ...
+    action: <Enforce or Warn>
+status:
+  conditions:
+    # Conditions are detailed further below
+    ...
+
+  # observedCRD is set only if the target CRD was observed.
+  observedCRD:
+    uid: <uid of the observed target CRD>
+    generation: <generation of the observed target CRD>
+
+```
+
+#### Conditions
+
+CRDCompatibilityRequirement defines 3 conditions:
+
+* `Progressing` -
+False if the spec has been completely reconciled.
+True indicates that reconciliation is still in progress and the current status does not represent a stable state.
+Progressing False with an error reason indicates that the object cannot be reconciled.
```

Comment:
Does this would imply that this immediately goes to `True` every time the spec is written? When would a controller write this to `True` in normal operation? While it's waiting for a webhook to be set up?